### PR TITLE
How to write tests section for Substrate Kitties

### PR DESCRIPTION
**Pull request by nczhu**
**Ready for review**

For #51 . Shout out to @Kianenigma for crux of 5.1-5.2. 

Concepts addressed include: 
- Section 5.1: setting up the mock for unit tests in substrate, which dependencies to import, how to impl traits for modules, test_externalities

- Section 5.2: writing unit tests for substrate, best practice around ensures, commonly used macros

- Section 5.3: updating test mock to set up genesis configs using build/with_storage. Demonstrates a new way to configure genesis not covered yet in other tutorials.

### CODE REVIEW

**Comment by Kianenigma** [comment]
One final idea that i had and forgot probably to put it anywhere:
not separating the tests into a separate `test.rs` file is
- Better in the sense that the things that the reader sees on the sidebar is less
- The code is not there for them too look at and write tests upon.

I don't have a strong preference.

For the rest of it, I am excited to give it a full read tomorrow. Thanks @nczhu.

**Comment by nczhu** [comment]
New section ready for review @Kianenigma

File: 5/assets/5.1-finished-code.rs [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+		type Hashing = BlakeTwo256;
+		type Digest = Digest;
+		type AccountId = u64;
+		type Lookup = IdentityLookup<Self::AccountId>;
+		type Header = Header;
+		type Event = ();
+		type Log = DigestItem;
+	}
+	
+    // And any other trait that your Trait is explicitly bounded by.
```
Comment (Kianenigma) [inline 5/assets/5.1-finished-code.rs]: oops

File: 5/assets/5.1-finished-code.rs [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+
+	// And finally, your own trait.
+	impl super::Trait for KittiesTest {
+		type Event = ();
+	}
+
+	// Creates aliases for modules for easier access
+	type Kitties = super::Module<KittiesTest>;
+	type Balances = balances::Module<KittiesTest>;
+	type System = system::Module<KittiesTest>;
```
Comment (Kianenigma) [inline 5/assets/5.1-finished-code.rs]: So as expected this is not really used :D neither is the balances Trait. We can either:
- Let them be as a reference for how to do it. I don't like the compiler warnings though.
- Even better, on the first `ACTION: write some kind of an assertion`, we can also ask them to do one assertion from balances and system as well, or just play around with them.

File: 5/assets/5.1-finished-code.rs [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+
+	fn build_ext() -> TestExternalities<Blake2Hasher> {
+		let mut t = system::GenesisConfig::<KittiesTest>::default().build_storage().unwrap().0;
+		t.extend(balances::GenesisConfig::<KittiesTest>::default().build_storage().unwrap().0);
+		// t.extend(GenesisConfig::<KittiesTest>::default().build_ext().unwrap().0);
```
Comment (Kianenigma) [inline 5/assets/5.1-finished-code.rs]: oops again.

File: 5/introduction.md [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+===
+
+At this point you have learned the essentials of Substrate runtime development and how to bootstrap a UI for your runtime.
+
+This next section will cover writing tests for your Substrate runtime. Tests are the backbone of any software artifact, ensuring its veracity as your codebase evolves. There are many _point-of-views_ when it comes to writing tests. We are not really focused on that on that in the scope of this tutorial. Rather, we will focus on the technical details of writing tests in Substrate.
+
+While you can test some of your code logic through the UI, it is best practice to write unit tests for each runtime module. Unit tests should cover the core functions of your runtime.
+
+Conveniently, Substrate has some low-level primitives that help you execute your runtime tests with minimum hassle, meanwhile giving it capabilities such as calling other modules and processing blocks to be as realistic as possible.
+
```
Comment (Kianenigma) [inline 5/introduction.md]: I probably wrote this because I had in mind to use `System::` somewhere.

File: 5/setting-up-tests.md [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+
+We start by creating a `test` module inside `substratekitties.rs` as follows. 
+
+> **Note:** Alternatively, you can also refactor this out into its own `test.rs` file.
+
+**substratekitties<span>.</span>rs**
+```rust
+// Your substratekitties code 
+
+#[cfg(test)]
```
Comment (Kianenigma) [inline 5/setting-up-tests.md]: We could write an appendix note about what this means.

File: 5/setting-up-tests.md [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+impl_outer_origin! {
+	pub enum Origin for KittiesTest {}
+}
+```
+
+## Construct a Mock Runtime
+
+Now we are ready to construct a mock runtime for our tests. 
+
+First, we declare a main configuration type `KittiesTest`. `KittiesTest` will implement each of the configuration traits of modules used by the Kitties runtime, such as `system` and 	`balances`.
```
Comment (Kianenigma) [inline 5/setting-up-tests.md]: extra tab.

File: 5/writing-tests.md [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+
+- Notice how using a `u64` as `AccountId` is handy. Your account is simply referred to as `10` rather than a special type.
+- Notice how we've used the same `Kitties` to call a getter function (e.g. `kitty_of_owner_by_index`) and a normal dispatch function (e.g. `create_kitty` from `decl_module!`).
+- Notice that you can use `Kitties` to call even the module functions as well (anything inside `impl Module<T>`)! Hence, you could (and should) have a test in which you only test the internal `mint()` function that is not directly accessible.
+
+> **Note:** When accessing storage, you can also use the following alternative syntax in lieu of `Kitties::get()`: 
+
+```rust
+use super::KittyOwner;
+use support::StorageMap;
```
Comment (Kianenigma) [inline 5/writing-tests.md]: Though this depends on the type of storage, so I would make it `use support::{StorageMap, StorageValue, DoubleMap, EnumerableStroageMap}` etc or mention that it __might__ be different.

File: 5/writing-tests.md [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+assert_eq!(<KittyOwner<KittiesTest>>::get(hash), Some(10));
+```
+
+## Recommended Test Pattern
+
+In Substrate, it is good practice to have comprehensive test coverage around your key state transition functions.
+
+Recall that Substrate provides an `ensure!` macro used to check inputs and logic before updating runtime state. In particular, whenever this macro is used, you should implement thorough tests around any edge cases that may break the ensured assumptions.
```
Comment (Kianenigma) [inline 5/writing-tests.md]: awesome. :100:

**Comment by Kianenigma** [comment]
I had some comments but all in all I think this is a great start. Next, we could also add another set of unit tests for each chapter. 

What I don't like is that each test template is containing the entire copy of the code, which is a bit hard to maintain as well. Shan't we just mention that this part is built on top of chapter X of this workshop and omit that copy-and-pasted code?

File: 5/introduction.md (hunk 4 above)
Comment (Kianenigma) [reply 5/introduction.md]: As in: in a dummy assertion that either we either write ourself or ask the user:
- set block number to X and read it from `System`.
- Set balance of X to Y and read it from `Balances`.

File: 5/testing-genesis.md [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+- Mocking the genesis configuration in during testing
+- Testing that we have the correct genesis set up
+
+## Add Extra Genesis
+
+Let's start by importing some functions and types we'll be using to configure genesis.
+
+**substratekitties<span>.</span>rs**
+```rust
+use runtime_io::{with_storage, StorageOverlay, ChildrenStorageOverlay};
```
Comment (shawntabrizi) [inline 5/testing-genesis.md]: Getting this error when compiling with Substrate v1.0:

```
error[E0432]: unresolved imports `runtime_io::with_storage`, `runtime_io::StorageOverlay`, `runtime_io::ChildrenStorageOverlay`
 --> /Users/shawntabrizi/Documents/GitHub/substrate-package/substrate-node-template/runtime/src/template.rs:8:18
  |
8 | use runtime_io::{with_storage, StorageOverlay, ChildrenStorageOverlay};
  |                  ^^^^^^^^^^^^  ^^^^^^^^^^^^^^  ^^^^^^^^^^^^^^^^^^^^^^ no `ChildrenStorageOverlay` in the root
  |                  |             |
  |                  |             no `StorageOverlay` in the root
  |                  no `with_storage` in the root
  |                  help: a similar name exists in the module: `set_storage`
```

Comment (nczhu) [reply 5/testing-genesis.md]: I have a working version in this branch: https://github.com/paritytech/substrate/tree/kiz-document-kitty. I will try to recreate it with substrate-package and see if I get the same issue.

Comment (shawntabrizi) [reply 5/testing-genesis.md]: So this probably uses master branch right?

Switch to v1.0 and you will likely run into the same issue. Possibly we need to backport whatever changes cause issues to the v1.0 branch, maybe open an issue?

File: 5/testing-genesis.md [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+The `config` attribute will expect a `kitties` config value, which will be a vector of tuples with types: 
+- AccountId: the kitty owner
+- Hash: the kitty id/dna
+- Balance: the kitty's initial value
+
+`build` subsequently uses the `kitties` configuration to build the storage itself.
+In this implementation, it iterates over `kitties` and use the `mint` function to conveniently update
+the rest of the storage items.
+
+> **Note**: For this part to work, delete (or comment out) the deposit_event line in `mint()`, as `add_extra_genesis`
```
Comment (shawntabrizi) [inline 5/testing-genesis.md]: This is a strange requirement, and ultimately not very helpful for someone who wants to write tests using this pattern... what are the solutions?

Comment (shawntabrizi) [reply 5/testing-genesis.md]: I was expecting something like this: https://github.com/paritytech/substrate/blob/master/srml/staking/src/mock.rs#L124

Comment (shawntabrizi) [reply 5/testing-genesis.md]: TBH I do not know much about `add_extra_genesis` and this section does not explain it well

**Comment by shawntabrizi** [comment]
Def an awesome pull request. Once things are working end to end, will def pull it in.

Room for improvement:

I think that we could expand a bit more explaining some of the new things we use in this section.

* Starting with the layout of: Mock, Test, Runtime

* Going more into details that Mock can use all these "psudo-types" because substrate is so generically written!

* Going deeper into overall genesis config (or using ext_builder, and saving genesis for something else)

* less "copy paste code", i.e. make the examples work on their own example logic, and then have the user really write their own code, rather than pasting from the documentation the right answers (not sure how well this will work, would be open to be told that this does not really apply here)

File: 5/writing-tests.md [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+#[test]
+fn it_works() {
+	with_externalities(&mut build_ext(), || {
+		assert!(true);
+	})
+}
+```
+
+The command for executing this test is: 
+`cargo test substratekitties.rs` or if want to test your entire package, simply run `cargo test -p YOUR_PACKAGE_NAME`.
```
Comment (shawntabrizi) [inline 5/writing-tests.md]: `cargo test substratekitties.rs` did not work for me.

`cargo test -p template-node-runtime` or whatever is what worked for me....

Is there some additional data we need to share to drive consistency here?

File: 5/testing-genesis.md (hunk 10 above)
Comment (nczhu) [reply 5/testing-genesis.md]: So we have 3 paths forward for this section: 

1. Marking storage values with `config()`, then configuring each storage value 1 by 1, as @shawntabrizi indicated. This is easier to understand and we have a few tutorials that already cover this. e.g. [1](https://docs.substrate.dev/docs/building-the-substrate-tcr-runtime#section-storage-declaration-using-the-decl_storage-macro) and [2](https://substrate.recipes/testing/index.html)

2. Building with build() inside each storage item declaration. I implemented this as well, before discovering the 3rd approach. i.e.
```rust
        Kitties get(kitty) build(|config: &GenesisConfig<T>| {
            config.kitties.iter()
                .map(|i| ( i.1, Kitty { id: i.1, dna: i.1, price: i.2, gen: 0, } ))
                .collect::<Vec<_>>()
        }): map T::Hash => Kitty<T::Hash, T::Balance>;

        KittyOwner get(owner_of) build(|config: &GenesisConfig<T>| {
            config.kitties.iter().cloned() //address this
                .map(|i| (i.1, i.0) )
                .collect::<Vec<_>>()
        }): map T::Hash => Option<T::AccountId>;
```

3. Waiting to backport latest master features to V1.0, so we can build storage inside `add_extra_genesis` (given a much better explanation 😄 ). 

I still advocate for 3 since Kitties' storage values are interdependent and this feels more elegant, less manual, less lines. 

But I do see value in also teaching (1) and (2), which are common patterns as well. Thoughts @shawntabrizi and @kianenigma ?

File: 5/writing-tests.md (hunk 11 above)
Comment (nczhu) [reply 5/writing-tests.md]: On second thought, the correct way we should teach them is: 
`cargo test -p substratekitties-runtime substratekitties`
Which reads as:  run only the substratekitties unit test inside the substratekitties-runtime subpackage. 

My understanding of this is that we have to run the test from root directory in order to satisfy dependencies. But in root, `substratekitties-runtime` is a subpackage, so has to be specified. which is why in this particular case, we cannot just do `cargo test substratekitties.rs`

**Comment by shawntabrizi** [comment]
Things are compiling for me now, only issue I see is the last "solution" fails to pass the tests:

```
running 4 tests
test template::tests::should_build_genesis_kitties ... ok
test template::tests::create_kitty_should_work ... FAILED
test template::tests::transfer_kitty_should_work ... FAILED
test template::tests::transfer_not_owned_kitty_should_fail ... ok

failures:

---- template::tests::create_kitty_should_work stdout ----
thread 'template::tests::create_kitty_should_work' panicked at 'assertion failed: `(left == right)`
  left: `3`,
 right: `1`', runtime/src/template.rs:327:13
note: Run with `RUST_BACKTRACE=1` environment variable to display a backtrace.

---- template::tests::transfer_kitty_should_work stdout ----
thread 'template::tests::transfer_kitty_should_work' panicked at 'assertion failed: `(left == right)`
  left: `2`,
 right: `1`', runtime/src/template.rs:359:13


failures:
    template::tests::create_kitty_should_work
    template::tests::transfer_kitty_should_work

test result: FAILED. 2 passed; 2 failed; 0 ignored; 0 measured; 0 filtered out
```

Other than that, I think we can merge in. @nczhu can you make sure that the tests are correct?

**Comment by nczhu** [comment]
@shawntabrizi : ready for final remeew.

Contains: 
- tests are fixed and passing
- better write up around add_extra_genesis
- content around other methods of setting up genesis/configs (not just this method)

**Comment by shawntabrizi** [comment]
💯
